File: packages/rust/genesis-ode/src/genesis_governance/multiplicity/allocator.py
from typing import List, Dict, Any, Optional
import numpy as np
from genesis_governance.shared.history import HistoryStore

class PrimeBandAllocator:
    """
    Automated prime-band allocation logic (ADR-009).
    Assigns primes to features based on sensitivity and historical performance.
    """
    
    BANDS = {
        "A": [2, 3, 5],     # Dynamics (Continuous)
        "B": [7, 11],       # Structure (Categorical)
        "C": [13, 17, 19]   # Context (Coupling/Parameters)
    }

    FEATURE_BAND_MAPPING = {
        "coherence": "A",
        "effective_stress": "A",
        "threshold_state": "A",
        "logical_state": "B",
        "frequency": "C",
        "coupling": "C"
    }

    def __init__(self, history: Optional[HistoryStore] = None):
        self.history = history
        # Default/Initial mapping
        self.current_mapping = {
            "coherence": 2,
            "effective_stress": 3,
            "threshold_state": 5,
            "logical_state": 7,
            "frequency": 13,
            "coupling": 17
        }

    def analyze_feature_sensitivity(self, feature_name: str) -> float:
        """
        Heuristic for feature sensitivity (0 to 1).
        In v0.5.0, we prioritize categorical states and high-variance features.
        """
        if feature_name == "logical_state":
            return 0.9
        
        if not self.history:
            return 0.5
            
        # Entropy heuristic: if we have history, look at variance in reconstruction scores
        recon_history = self.history.get_reconstruction_history()
        target_scores = [r["score"] for r in recon_history if r["target_id"] == feature_name]
        
        if len(target_scores) > 10:
            return float(np.std(target_scores))
            
        return 0.4

    def allocate(self) -> Dict[str, int]:
        """
        Updates and returns the feature-to-prime mapping.
        """
        return self.current_mapping

    def sensitivity_sweep(self, score_threshold: float = 0.80, delta_threshold: float = 0.15) -> bool:
        """
        Queries HistoryStore for reconstruction drift.
        Re-allocates if drift exceeds thresholds. Returns True if re-allocation occurred.
        """
        if not self.history:
            return False
            
        recon_history = self.history.get_reconstruction_history()
        if not recon_history:
            return False
            
        # Analyze latest window
        window = recon_history[-50:]
        avg_score = np.mean([r["score"] for r in window])
        max_delta = np.max([r["delta"] for r in window]) if window else 0.0
        
        reallocated = False
        if avg_score < score_threshold or max_delta > delta_threshold:
            # TRIGGER RE-ALLOCATION: Cycle primes within their bands
            for feature, band_id in self.FEATURE_BAND_MAPPING.items():
                primes_in_band = self.BANDS[band_id]
                current_p = self.current_mapping[feature]
                
                # Simple rotation within the band
                current_idx = primes_in_band.index(current_p)
                next_p = primes_in_band[(current_idx + 1) % len(primes_in_band)]
                
                if next_p != current_p:
                    self.current_mapping[feature] = next_p
                    reallocated = True
                
        return reallocated

    def get_prime_for_feature(self, feature_name: str) -> int:
        return self.current_mapping.get(feature_name, 23) # fallback to high prime

File: packages/rust/genesis-multiplicity/src/allocator.rs
//! Port of `genesis_governance/multiplicity/allocator.py` — automated
//! prime-band allocation (ADR-009).

use crate::history::{HistoryStore, ReconstructionRecord};
use std::collections::HashMap;

/// Prime bands: `BANDS` from the Python (dynamics / structure / context).
pub const BANDS: &[(&str, &[i64])] = &[
    ("A", &[2, 3, 5]),    // Dynamics (Continuous)
    ("B", &[7, 11]),      // Structure (Categorical)
    ("C", &[13, 17, 19]), // Context (Coupling/Parameters)
];

/// Feature → band assignment, insertion order preserved (the sweep rotates
/// features in this exact order, matching the Python dict iteration order).
pub const FEATURE_BAND_MAPPING: &[(&str, &str)] = &[
    ("coherence", "A"),
    ("effective_stress", "A"),
    ("threshold_state", "A"),
    ("logical_state", "B"),
    ("frequency", "C"),
    ("coupling", "C"),
];

/// Default/initial feature-to-prime mapping, mirroring `current_mapping`.
pub fn default_mapping() -> HashMap<String, i64> {
    HashMap::from([
        (String::from("coherence"), 2),
        (String::from("effective_stress"), 3),
        (String::from("threshold_state"), 5),
        (String::from("logical_state"), 7),
        (String::from("frequency"), 13),
        (String::from("coupling"), 17),
    ])
}

fn band(band_id: &str) -> &'static [i64] {
    BANDS
        .iter()
        .find(|(id, _)| *id == band_id)
        .map(|(_, primes)| *primes)
        .unwrap_or(&[])
}

/// Automated prime-band allocation logic (ADR-009).
#[derive(Debug, Clone)]
pub struct PrimeBandAllocator {
    pub history: Option<HistoryStore>,
    pub current_mapping: HashMap<String, i64>,
}

impl PrimeBandAllocator {
    pub fn new(history: Option<HistoryStore>) -> Self {
        Self {
            history,
            current_mapping: default_mapping(),
        }
    }

    /// Heuristic for feature sensitivity (0 to 1). Mirrors the Python
    /// exactly: `logical_state` is prioritized, missing history yields `0.5`,
    /// and >10 targeted history samples yield the population std-dev.
    pub fn analyze_feature_sensitivity(&self, feature_name: &str) -> f64 {
        if feature_name == "logical_state" {
            return 0.9;
        }
        let Some(history) = &self.history else {
            return 0.5;
        };
        let recon_history = history.get_reconstruction_history();
        let target_scores = recon_history
            .iter()
            .filter(|r| r.target_id.as_deref() == Some(feature_name))
            .filter_map(|r| r.score)
            .collect::<Vec<_>>();
        if target_scores.len() > 10 {
            return crate::population_std(&target_scores);
        }
        0.4
    }

    /// Returns the current feature-to-prime mapping (a snapshot; the Python
    /// returns the live dict, so a fresh allocator is the correct analogue).
    pub fn allocate(&self) -> HashMap<String, i64> {
        self.current_mapping.clone()
    }

    /// Queries `HistoryStore` for reconstruction drift and re-allocates
    /// (rotates within bands) when drift exceeds the thresholds. Returns
    /// `true` when any feature moved.
    pub fn sensitivity_sweep(&mut self, score_threshold: f64, delta_threshold: f64) -> bool {
        let Some(history) = &self.history else {
            return false;
        };
        let recon_history = history.get_reconstruction_history();
        if recon_history.is_empty() {
            return false;
        }

        // Analyze latest window: Python `recon_history[-50:]`.
        let window = &recon_history[recon_history.len().saturating_sub(50)..];
        let avg_score = mean(window);
        let max_delta = window
            .iter()
            .map(|r| r.delta.unwrap_or(0.0))
            .fold(0.0_f64, f64::max);

        let mut reallocated = false;
        if avg_score < score_threshold || max_delta > delta_threshold {
            // TRIGGER RE-ALLOCATION: cycle primes within their bands.
            for &(feature, band_id) in FEATURE_BAND_MAPPING {
                let primes_in_band = band(band_id);
                let current_p = self.current_mapping[feature];
                let current_idx = primes_in_band
                    .iter()
                    .position(|&p| p == current_p)
                    .unwrap_or(0);
                let next_p = primes_in_band[(current_idx + 1) % primes_in_band.len()];
                if next_p != current_p {
                    self.current_mapping.insert(feature.to_string(), next_p);
                    reallocated = true;
                }
            }
        }
        reallocated
    }

    /// `mapping.get(feature, 23)` — fallback to a high prime.
    pub fn get_prime_for_feature(&self, feature_name: &str) -> i64 {
        self.current_mapping
            .get(feature_name)
            .copied()
            .unwrap_or(23)
    }
}

fn mean(records: &[ReconstructionRecord]) -> f64 {
    if records.is_empty() {
        return 0.0;
    }
    records.iter().map(|r| r.score.unwrap_or(0.0)).sum::<f64>() / records.len() as f64
}
